node-discovery/src/Ergvein/Node/Parsing.hs

```diff
@@ -0,0 +1,93 @@
+module Ergvein.Node.Parsing
+  ( NamedSockAddr(..)
+  , parseSockAddr
+  , parseSockAddrs
+  , parseHostPort
+  , ipAddressParser
+  ) where
+
+import Data.Text (Text)
+import Network.Socket (SockAddr)
+import Control.Monad
+import Control.Applicative
+import Control.Monad.IO.Class
+import Data.Either
+import Data.IP
+import Data.Maybe
+import Network.DNS.Lookup
+import Network.DNS.Resolver
+import Network.DNS.Types
+import Network.Socket
+import Text.Read (readMaybe)
+import qualified Data.List.Safe as LS
+import qualified Data.Text as T
+import qualified Data.ByteString.Char8 as B8
+import Data.Attoparsec.Text
+import Data.Word
+
+data NamedSockAddr = NamedSockAddr {
+  namedAddrName :: Text
+, namedAddrSock :: SockAddr
+} deriving (Eq, Ord)
+
+parseSockAddrs :: (MonadIO m) => ResolvSeed -> PortNumber -> [Text] -> m [NamedSockAddr]
+parseSockAddrs rs defNodePort urls = liftIO $ catMaybes <$> (withResolver rs $ \r -> forM urls $ parseAddr r defNodePort)
+
+parseSockAddr :: (MonadIO m) => ResolvSeed -> PortNumber -> Text -> m (Maybe NamedSockAddr)
+parseSockAddr rs defNodePort t = liftIO $ withResolver rs $ \r -> parseAddr r defNodePort t 
+
+parseAddr :: Resolver -> PortNumber -> Text -> IO (Maybe NamedSockAddr)
+parseAddr resolver defNodePort addressText =
+  case parseOnly (ipAddressParser defNodePort) addressText of 
+    Right ip -> pure $ Just $ NamedSockAddr addressText ip
+    _-> do
+      let (domain, port) = parseHostPort defNodePort addressText
+      v4 <- lookupA resolver domain
+      v6 <- lookupAAAA resolver domain
```

Comment:
The code always perform two lookups even one is enough. Lookup ipv6 address only if there is no ipv4 address available.